### DISCUSSION TITLE
Production ready?

### DISCUSSION THREAD

**Discussion by cloudcompute** [wasmCloud/wasmCloud#270] [Q&A] [answered]
Any one using it in production? How can we integrate it with Keycloak IAM, need to write its capability provider?

**Comment by thomastaylor312** [comment]
Hi there @cloudcompute! Thanks for the question. So as for production worthiness, at Cosmonic (my current job), the entire product is built on top of wasmcloud. Also there is a use case blog from Adobe: http://cncf.io/blog/2022/11/17/better-together-a-kubernetes-and-wasm-case-study/ and various data points from people using it: https://www.trustradius.com/products/wasmcloud/reviews. We are also planning on moving towards Incubating status in the CNCF this year.

As for your second question, for integration with Keycloak you would need to write its capability provider. That is actually the answer to most integration questions! Let us know if you need any guidance on how to do that

**Comment by cloudcompute** [comment]
Hi @thomastaylor312 

Thanks for the reply.

> at Cosmonic (my current job), the entire product is built on top of wasmcloud.

Great !

I am looking forward to write an event-driven, event-sourced microservices-based application. Its architecture is:
- Client apps write events to an Event store (two options: either SQL database Or Nats JetStream itself)
- If SQL db, push the events to Nats for pub-sub
- At the other, different subscribers consume the events, process them, and update their views/projections (their respective data stores)

How does WasmCloud fit in this scenorio, I have several queries relating to it. I hardly know Wasm, so some questions of mine are very basic, so kindly accommodate.

a. Could the `actors` be used for `long running processes` (micro-services) or are they simply `serverless`?

b. If Actors are intended to serve serverless workloads only, then how can we take care of long running processes while using WasmCloud? Do we need Kubernetes for their auto scaling, discovery etc.? What about the database processes, they won't be a part of the WasmCloud, right?

c. I guess these actors simply executables and they don't need any runtime.. is this the reason that `we don't need containers` when using WasmCloud?

d. In which `language can we write actors`, Rust and Go only.. Or, can we use any other language like Java, Kotlin, or JavaScript?

e. I think I can use `Nats Jetstream` for event-sourcing which is `already a part of WasmCloud`? Does it auto-scale out-of-box, or we need to ensure its scalability ourselves? By the way, would you recommend using Nats as an event source (if you have come across any such implementation), or shall we use a dedicated event store like PostgreSQL?

f. Is there `really a need to use Kubernetes` when writing apps on top of WasmCloud? When using WasmCloud, do we really need Kubernetes for features like fault-tolerance, service discovery? I think we don't need something like horizontal pod autoscalers, as actors are autoscalable.

The reason you guys have chosen Elixir to write wasmcloud-otp is that its underlying VM has excellent features like fault-tolerant, distributed nature, etc. And is this what helping us stay away from Kubernetes?

g. How does WasmCloud exactly `take care of NFAs`. They are numerous. At one place, I read there is `no need for explicit firewalls`, Wasm takes care of it. Does this mean Web Application Firewall that has features like DDoS?

h. Is there any `detailed documentation` available? The current docs has little information for newbies like me.

> for integration with Keycloak ... Let us know if you need any guidance on how to do that

I'd seek your help if I decide to use WasmCloud.

Thank you!

**Comment by thomastaylor312** [comment, marked as answer]
Those are a lot of good questions. I think the best way to help with all of those would be to join [our slack](https://slack.wasmcloud.com/) and ask there!

As for documentation, you can find it at https://wasmcloud.dev. For capability providers in particular: https://wasmcloud.dev/app-dev/create-provider/.

It is also cool to hear that you would like to do an event sourcing model on top of it. Turns out, this is what we did at Cosmonic!. We're hoping to open source some of that in the next few months, but can definitely provide guidance with more specific details in the wasmCloud Slack!

**Comment by brooksmtownsend** [comment]
Hey @cloudcompute, with the [1.0.0 release](https://github.com/wasmCloud/wasmCloud/releases/tag/v1.0.0) you may be interested in checking out the extra stability guarantees and updated documentation on https://wasmcloud.com 😄

**Comment by cloudcompute** [comment]
Hi @brooksmtownsend 

A fortnight ago, I had come to your site to check about the progress. Great to know that 1.1 is there. 

Thank you for letting me know.
